### ISSUE TITLE
pytorch not respecting torch.set_num_threads

### ISSUE THREAD

**Issue by tudor-berariu** [pytorch/pytorch#975]
In a recent update we observed an important performance fall in the backward phase on all our projects.

We used this [benchmark](https://github.com/bit-ml/tensor-benchmarks/blob/master/nn_benchmark.py) to measure the computation times for the forward and the backward phases in a simple neural model.

In summary we observed that the oldest commit with good performance is @761d679. The following commit (@65b6626) introduces a huge amount of extra time during error backpropagation. The current performance (which we belive dates from commit @15a9fbd) is a bit better than that, but still worse than @761d679.

Here are some running times we got with our benchmark on a Lenovo Y50 with Intel i7-4710HQ CPU @ 2.50GHz and nVidia GeForce 860M. We got similar results on GeForce 960M and Titan X (not shown here). (edited)

- commit @761d679
```
---------- FORWARD ----------
FORWARD @ CPU ===> 4.0940 s.
FORWARD @ GPU ===> 1.2921 s.
---------- BACKWARD ----------
BACKWARD @ CPU ===> 9.2027 s.
BACKWARD @ GPU ===> 2.1065 s.
```
- commit @65b6626
```
---------- FORWARD ----------
FORWARD @ CPU ===> 13.5454 s.
FORWARD @ GPU ===> 1.3190 s.
---------- BACKWARD ----------
BACKWARD @ CPU ===> 22.9335 s.
BACKWARD @ GPU ===> 2.1490 s.
```

- commit @15a9fbd
```
---------- FORWARD ----------
FORWARD @ CPU ===> 4.1922 s.
FORWARD @ GPU ===> 1.2867 s.
---------- BACKWARD ----------
BACKWARD @ CPU ===> 15.0562 s.
BACKWARD @ GPU ===> 2.2064 s.
```

- latest version (performance is similar with @15a9fbd)
```
---------- FORWARD ----------
FORWARD @ CPU ===> 4.2619 s.
FORWARD @ GPU ===> 1.2833 s.
---------- BACKWARD ----------
BACKWARD @ CPU ===> 16.4800 s.
BACKWARD @ GPU ===> 2.0648 s.
```

Is this performance trade-off expected and accounted for?

**Comment by soumith** [comment]
your issue and investigation is impeccable, thank you :)
We're looking into it and will reply back

**Comment by mattmacy** [comment]
@soumith could someone at FAIR setup a Jenkins to do basic performance collection after each commit?

**Comment by ngimel** [comment]
You are not doing torch.cuda.synchronize, so your GPU timings are likely off. The fact that CPU timings changed after @65b6626 is incredibly strange, as it concerns GPU only. Likely also due to lack of synchronization, because CPU and GPU performance is measured in the same loop. For that matter, the fact that even GPU performance changed after @65b6626 is also strange, because it concerns multi-GPU only, and your benchmark is single GPU.

**Comment by albanD** [comment]
@tudor-berariu I think your GPU backward timings are wrong due to a typo [here](https://github.com/bit-ml/tensor-benchmarks/blob/219fffee72b7493cd14bdc2890e92be601c83a05/nn_benchmark.py) it should be `total` and not `t`.
If you compute the difference by hand between the GPU and CPU time, you have a backward that is always ~1.5s.

The variance on the CPU runtime is quite weird though.

**Comment by tudor-berariu** [comment]
I also corrected the script fixing the two errors you pointed out. I'll run the tests again. Thank you!

**Comment by tudor-berariu** [comment]
@albanD It seems that @761d679 uses just 4 threads, while in more recent commits (@65b6626 or @15a9fbd) use more threads during backward phase. You were right, there's no problem when executing on the GPU.

I set the number of threads to 4.    `torch.set_num_threads(4)`

**Comment by tudor-berariu** [comment]
The times on GPU are correct, our initial measurements were wrong. I am really sorry for that.

The difference in performance on CPU is caused by the fact that during the backward phase recent versions of pytorch (newer than @65b6626) use more than the number of OpenMP threads set programmatically.

**Comment by floringogianu** [comment]
We decided to look into the performance figures when the slowdown became obvious on a CPU-heavy experiment (Mnih's 2015 A3C) which does backprop steps on each different agent/process. Specifically commit @761d679 finishes playing 96,000 steps in **14 seconds**  while with the latest commit it finishes the same task in **175 seconds**!

**Comment by soumith** [comment]
To answer @mattmacy 's high level question, I've bought an additional machine to setup perf-regression contbuilds, I'll get that setup in the next week or two. we will additionally be setting up OSX contbuilds as well.

**Comment by apaszke** [comment]
Are you using MKL? You might need to set `MKL_NUM_THREADS` too

**Comment by floringogianu** [comment]
@apaszke yes, mkl and anaconda3. MKL sees the correct no of threads available in the system. The issue is that since the backward pass was parallelized a few commits ago, when launching multiple processes  'set_num_threads(1)' has no effect anymore and each process launched tries to use all the available threads for blas ops which in turn slows down everything to a halt. 

We realize now the benchmarks above are not really illustrative of the issue, we'll get back first thing tomorrow with a benchmark using 'torch.multiprocess'.

**Comment by apaszke** [comment]
The backward pass is parallelized only on a per-device basis, so if your graph is CPU only, it should still be dispatched using a single thread.

**Comment by tudor-berariu** [comment]
@apaszke Here is the CPU load when running the same [benchmark](https://github.com/bit-ml/tensor-benchmarks/blob/master/nn_benchmark.py) on the same machine with two separate fresh anaconda+pytorch installs with the latest commit and, respectively, commit @761d679. The screens are captured during the backward phase on CPU.
The latest commit seems to ignore `torch.set_num_threads(4)` during error backpropagation.

![latest](https://cloud.githubusercontent.com/assets/1908458/23822819/2dcd3f88-065d-11e7-9e17-4a1441b399b1.png)
![761d679](https://cloud.githubusercontent.com/assets/1908458/23822821/382442ce-065d-11e7-8973-1ee33ac44506.png)

**Comment by apaszke** [comment]
What if you try setting both `OMP_NUM_THREADS` and `MKL_NUM_THREADS`? Does it help?

**Comment by apaszke** [comment]
Setting `OMP_NUM_THREADS` seems to help. I've looked into the issue and it seems that the problem is caused by `zero` calls in [`mm`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/generic/methods/TensorMath.cwrap#L1445) and [`mv`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/generic/methods/TensorMath.cwrap#L1424). Removing them speeds up the backward from 15.6s to 4.4s (forward takes 3.5s), without touching `OMP` or `MKL_NUM_THREADS` (so using all 40 cores).

A recent PR that added support for AVX instructions made `fill` use OpenMP, and this is why all these threads are spawned, but I'm not sure why does it slow down everything so much. Each of them is pre-assigned a single contiguous chunk of size `numel / omp_num_threads` to process, so they shouldn't be thrashing the cache (maybe it's just too small for them?). What's more, if I time a call to `tensor.zero_()` in an interpreter, it never gets slower when I increase the number of threads (however the speedups are quite small after reaching 20 threads). @adamlerer any ideas what might be going on?

I think the problem with `set_num_threads` might have been caused by ff5fa11 that, as I understand, will additionally link Intel OpenMP. Since we're still going to be adding `-fopenmp` to the compile flags a guess would be that GNU OpenMP would still be linked, leading to a symbol conflict, so `omp_set_num_threads` gets resolved to whichever library is loaded first, and it makes it possible to control how many threads will be used only by MKL, or only by torch libs. Output of `ldd` seems to confirm that theory, and since `iomp5` appears before `gomop`, all omp symbols are resolved form it:

```
ldd torch/lib/libTH.so.1
    ...
    libiomp5.so => /.../miniconda3/lib/libiomp5.so (0x00007f10f15af000)
    ...
    libgomp.so.1 => /.../miniconda3/lib/libgomp.so.1 (0x00007f10f0cf7000)
...
```


To sum up I think there are two problems:
1. We're suffering from having two omp libraries linked at the same time
2. For some reason `zero` calls in our bindings are very very slow

**Comment by tudor-berariu** [comment]
Setting `OMP_NUM_THREADS` does indeed solve the performance gap for the single process benchmark.
We are still trying to make use of these variables in order to match the @761d679 performance for programs using multiprocess and doing most work on CPU. We are still experiencing ~2x slowdowns in those cases even when setting `OMP_NUM_THREADS`.

**Comment by apaszke** [comment]
That's weird, multiprocessing case shouldn't change a lot. Could you try to get us a benchmark that reproduces it?

**Comment by floringogianu** [comment]
This is the [benchmark](https://github.com/bit-ml/tensor-benchmarks/blob/master/mp_benchmark.py).

Also, the `ldd` output on one of our servers:
```
$ ldd torch/lib/libTH.so.1 
    linux-vdso.so.1 =>  (0x00007ffeb1fc9000)
    libmkl_intel_lp64.so => not found
    libmkl_intel_thread.so => not found
    libmkl_core.so => not found
    libiomp5.so => not found
    libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fd6795fb000)
    libm.so.6 => /lib64/libm.so.6 (0x00007fd6792f9000)
    librt.so.1 => /lib64/librt.so.1 (0x00007fd6790f1000)
    libgomp.so.1 => /lib64/libgomp.so.1 (0x00007fd678ed9000)
    libc.so.6 => /lib64/libc.so.6 (0x00007fd678b17000)
    /lib64/ld-linux-x86-64.so.2 (0x00007fd679dcf000)
```

**Comment by apaszke** [comment]
I can't reproduce it anymore. Running just `python mp_benchmark.py` takes 99s, but once I do `OMP_NUM_THREADS=1 python mp_benchmark.py` the time drops to 8.7s, which is exactly the same as if I ran using the 761d679 build.

**Comment by floringogianu** [comment]
In general we noticed @761d679 finishing 3x times faster than `OMP_NUM_THREADS=1 python mp_benchmark.py` on the latest commits, especially when launching more processes. So while setting the variable helps a lot, there is still something off happening.

We have pytorch installed on several machines and we get different performance figures. Right now we don't feel we can control the installation process well enough. For example on two identical machines commit @4647f753 performs completely different: machine A performs simillar with @761d679 and is our baseline. Machine B with @4647f753 is at best 3x slower (even with `OMP_NUM_THREADS`) and slows down to a halt wihout it. However it performs the same as machine B with @761d679. `ldd` has the same output as above in *all cases*.

Please let me know if there's anything else we can look into and report back. Not sure if these details are relevant for this issue anymore.

**Comment by apaszke** [comment]
Well, first thing would be to solve the linking problem properly. It seems that we might need to switch to clang if we want to use MKL (see [this SO thread](http://stackoverflow.com/questions/25986091/telling-gcc-to-not-link-libgomp-so-it-links-libiomp5-instead)).

I have no idea why are you seeing different results on these machines.

**Comment by ngimel** [comment]
ff5fa11 was awful (I made a disclaimer that I don't know nearly enough about compiler/linker/threading library interactions) but it came about because conda mkl package that is used in pytorch conda packages and in many source installs does not come with libmkl_gnu_thread. Rather than switching to clang may be it is better to use mkl distribunions that come with libmkl_gnu_thread and revert ff5fa11.

**Comment by colesbury** [comment]
It looks like the anaconda/mkl pacakge doesn't include `libmkl_gnu_thread`, but the `intel/mkl` package does. We could switch to that, but that forces a few other packages (like pillow) to downgrade since they seem to depend on anaconda/mkl.

**Comment by colesbury** [comment]
Actually, it looks like things work with `libiomp5.so`. I've tested the latest PyTorch on a clean Anaconda installation on EC2's Ubunut 16.04 instance. Both forward and backward use the correct number of OpenMP threads.

After @ngimel's change, MKL now uses more than one thread by default. OpenMP by default uses the same number of threads as CPUs which is often too many and slows things down. Perhaps this is what you saw.

The other thing is that `torch.set_num_threads()` only affects the current process. If you launch a child process, then you have to call `torch.set_num_threads()` in that process as well. (I see you're doing this in mp_benchmark.py) The environment variable `OMP_NUM_THREADS` is inherited by child processes, like other environment variables.

**Comment by floringogianu** [comment]
> Actually, it looks like things work with libiomp5.so. I've tested the latest PyTorch on a clean Anaconda installation on EC2's Ubunut 16.04 instance. Both forward and backward use the correct number of OpenMP threads.

Is there anything besides `export CMAKE_PREFIX_PATH=[anaconda root directory]` you are doing to control the installation environment? Maybe you are also creating a conda virtual env? Because for some reasons `libTH.so.1` is not linking properly for me (see above.)

> After @ngimel's change, MKL now uses more than one thread by default. OpenMP by default uses the same number of threads as CPUs which is often too many and slows things down. Perhaps this is what you saw.

Yes, this is what I am seeing too. That on the backward pass torch launches more threads than optimal. And that in multiprocessing programs `set.num.threads()` has no effect. Setting `OMP_NUM_THREADS` helps but compared to @761d679 it's still about 3x still.

> The other thing is that torch.set_num_threads() only affects the current process. If you launch a child process, then you have to call torch.set_num_threads() in that process as well. (I see you're doing this in mp_benchmark.py) The environment variable OMP_NUM_THREADS is inherited by child processes, like other environment variables.

Yes, I'm careful to `set_num_threads()` on each process I launch. However with the most recent commits I also need to set `OMP_NUM_THREADS`. 

But what's more baffling is that for some reason @761d679 is the only recent commit that: 
- consistently builds properly, 
- uses the correct no of threads on the backward pass, 
- picks the correct no of OpenMP threads (without setting any flags) and
- offers the best performance figures in the benchmarks we posted and with the programs we are running.

This despite the fact that `ldd` shows that `libTH.so.1` links the same as in all other cases.

**Comment by floringogianu** [comment]
And just one more bit of info. The latest binaries are running fine now on our servers. So for now we have our performance issues resolved. It would be great to figure out how to build from sources properly though.

**Comment by liboyue** [comment]
Same problem here. I am using OSX Sierra 10.12.3, Anaconda 2.

```
$ otool -L libTH.1.dylib
libTH.1.dylib:
	@rpath/libTH.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libmkl_intel_lp64.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_intel_thread.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_core.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libiomp5.dylib (compatibility version 5.0.0, current version 5.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.0.0)
```
Setting ``MKL_NUM_THREADS`` and ```OMP_NUM_THREADS``` is useless.

**Comment by floringogianu** [comment]
Update regarding the binaries. The newer, `v0.1.11` build is now requiring the `OMP_NUM_THREADS` flag. But even with the flag the commits on March 6th are causing a **2x** slowdown compared to `v0.1.10` which uses a commit from March 5th.

Again that is on multiprocessing code, with forward, backward operations on multiple processes. The commits on March 6th are about parallelizing backward operations, I guess that's where the **2x** overhead coming...

**Comment by soumith** [comment]
multiple processes are running a CPU model forward+backward, correct? No GPU involved? 
And you are setting OMP_NUM_THREADS=1 and MKL_NUM_THREADS=1 ?
(i ask because multithreaded backward is only invoked if we have GPUs)

**Comment by floringogianu** [comment]
Yes, I am setting `OMP_NUM_THREADS=1` and `MKL_NUM_THREADS=1`. The first flag actually suffice. No GPU involved.

**Comment by soumith** [comment]
okay. do you have GPUs on your system? I see that just having GPUs is enough to invoke multithreaded backward.
If you do have GPUs on your system, can you also export `CUDA_VISIBLE_DEVICES=""`

**Comment by floringogianu** [comment]
Yes, I have GPUs. Setting `CUDA_VISIBLE_DEVICES=""` causes an error when calling `backward()` on a `Variable` however:
```
RuntimeError: cuda runtime error (38) : no CUDA-capable device is detected at torch/csrc/autograd/engine.cpp:353
RuntimeError: cuda runtime error (38) : no CUDA-capable device is detected at torch/csrc/autograd/engine.cpp:353
Traceback (most recent call last):
....
  File "mp_benchmark.py", line 40, in play
    loss.backward()
  File "/home/florin/Tools/anaconda3/lib/python3.6/site-packages/torch/autograd/variable.py", line 146, in backward
    self._execution_engine.run_backward((self,), (gradient,), retain_variables)
RuntimeError: cuda runtime error (38) : no CUDA-capable device is detected at torch/csrc/autograd/engine.cpp:353
```

**Comment by soumith** [comment]
ok, that was fixed in master but the error still exists in `0.1.11` i think.
Can you try:

`CUDA_VISIBLE_DEVICES=0`

**Comment by floringogianu** [comment]
No change, sorry.

**Comment by soumith** [comment]
thanks, will figure out what's wrong. I'll use the same benchmark code.

**Comment by soumith** [comment]
this PR by @colesbury should've fixed this issue guys. https://github.com/pytorch/pytorch/pull/1740

@tudor-berariu @floringogianu if any of you can verify, i'll close the issue.

**Comment by apaszke** [comment]
I don't think it's solved. We're still linking against both `libgomp` and `libiomp`, which means that without env variables, users can't reliably set the number of threads (depending on which one is loaded first, they can only control MKL threads or TH threads, but never both). It can also lead to some very weird behaviour in the future, because one of these libraries probably has its public symbols resolved to these from the other one. This is what `ldd` shows for me:
```
	libiomp5.so => /home/apaszke/miniconda3/lib/libiomp5.so (0x00007f277a9dc000)
	libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f277a7ba000)
```

**Comment by ttpro1995** [comment]
is this fixed yet ?

**Comment by soumith** [comment]
this was fixed for `0.3.1`, we also set `mkl_num_threads` when we do torch.set_num_threads. I think it can be closed.

**Comment by deepbrain** [comment]
in a build from the source snapshot as of today, I am still seeing the performance drop of 50% unless I set the OMP_NUM_THREADS=1

**Comment by fwpheckel** [comment]
I am also seeing that `set_num_threads` is not having the desired effect, and `OMP_NUM_THREADS=1` is the effective way to force single-threaded behavior.

**Comment by streamnsight** [comment]
I can also report that I see `set_num_thread` has no effect

Another 'funny' behavior I notice using the `OMP_NUM_THREADS` and `MKL_NUM_THREADS` variables, is that with my machine, a Dual Xeon, setting those to `N` ends up spinning `2N` threads except for N=1, so `OMP_NUM_THREADS=2` results in 4 threads, `OMP_NUM_THREADS=3` results in 6 threads etc...

I'm using a simple nn.Linear layer for testing and I don't see a drastic improvement using multiple threads. It actually seems to saturate at 12 threads (i.e. no further gain in speed) and the speed gain is definitely not linear (i.e. from 1->2 threads I get a 50% gain, 2->4 ~20%, 4->6, ~10%)

I'm confused about what I can expect as a gain from the multi processing capability. It seems to bring a LOT of overhead.
